code/contracts/ONEWallet.sol

```diff
@@ -64,30 +65,33 @@ contract ONEWallet {
         (uint32 ct, bool completed) = _findCommit(hash);
         require(ct == 0 && !completed, "Commit already exists");
         Commit memory nc = Commit(hash, uint32(block.timestamp), false);
-        commits.push(nc);
+        commits.push(nc); // IH: would not it be better to have mapping of commits?
     }
 
     // TODO (@polymorpher): during reveal, reject if index corresponds to a timestamp that is same as current block.timestamp, so we don't let the client accidentally leak EOTP that is still valid at the current time, which an attacker could use to commit and reveal a new transaction. This introduces a limitation that there would be a OTP_INTERVAL (=30 seconds by default) delay between commit and reveal, which translates to an unpleasant user experience. To fix this, we may slice OTP_INTERVAL further and use multiple EOTP per OTP_INTERVAL, and index the operations within an interval by an operation id. For a slice of size 16, we will need to generate a Merkle Tree 16 times as large. In this case, we would reject transactions only if both (1) the index corresponds to a timestamp that is same as current block.timestamp, and, (2) the operation id is not less than the id corresponding to the current timestamp's slice
+    // IH: what do you state above just supports HOTPs instead of TOTPs, besides other advantages.
+    // IH: combining index with nonce is unnecessary complicated to me. I would split them. 
     function revealTransfer(bytes32[] calldata neighbors, uint32 indexWithNonce, bytes32 eotp, address payable dest, uint256 amount) external
     isCorrectProof(neighbors, indexWithNonce, eotp)
     returns (bool)
     {
-        bytes memory packedNeighbors = _pack(neighbors);
+        bytes memory packedNeighbors = _pack(neighbors); // IH: no need to put Merkle proof into the hash (it is either correct and matches to OTP or not, hence integrity is implicit)
         bytes memory packed = bytes.concat(packedNeighbors,
             bytes32(bytes4(indexWithNonce)), eotp, bytes32(bytes20(address(dest))), bytes32(amount));
-        bytes32 commitHash = keccak256(bytes.concat(packed));
+        bytes32 commitHash = keccak256(bytes.concat(packed)); // IH: why do you need bytes.concat again?
         _revealPreCheck(commitHash, indexWithNonce);
         _completeReveal(commitHash);
-        uint32 day = uint32(block.timestamp) / 86400;
+        uint32 day = uint32(block.timestamp) / 86400; // IH: goof to use a constant; also casting uint 256 as uint32 does not seem reasonable (maybe if you'd put division into the parenthesis). But anyway, I'd go for uint instead.
         if (day > lastTransferDay) {
             spentToday = 0;
             lastTransferDay = day;
         }
-        if (spentToday + amount > dailyLimit) {
+        if (spentToday + amount > dailyLimit) { // IH: I'd put revert here. So the user will be warned by client that tx will not pass and do not proceed => save gas.
```

Comment:
We should not revert here, because it would undo the reveal (and nonce increment), making it possible to re-use the same OTP and its proof even though it is already revealed exposed.